htlcswitch/circuit.go

```diff
@@ -1,135 +1,170 @@
+// payment circuit. The circuit map is keyed with the idenitifer for the
+// outgoing HTLC
+type circuitKey struct {
+	chanID lnwire.ShortChannelID
+	htlcID uint64
 }
 
-// isEqual checks the equality of two payment circuits.
-func (a *paymentCircuit) isEqual(b *paymentCircuit) bool {
-	return bytes.Equal(a.PaymentHash[:], b.PaymentHash[:]) &&
-		a.Src == b.Src &&
-		a.Dest == b.Dest
+// String returns a string representation of the circuitKey.
+func (k *circuitKey) String() string {
+	return fmt.Sprintf("(Chan ID=%s, HTLC ID=%d)", k.chanID, k.htlcID)
 }
 
-// circuitMap is a data structure that implements thread safe storage of
-// circuits. Each circuit key (payment hash) may have several of circuits
-// corresponding to it due to the possibility of repeated payment hashes.
+// CircuitMap is a data structure that implements thread safe storage of
+// circuit routing information. The switch consults a circuit map to determine
+// where to forward HTLC update messages. Each circuit is stored with it's
+// outgoing HTLC as the primary key because, each offered HTLC has at most one
+// received HTLC, but there may be multiple offered or received HTLCs with the
+// same payment hash. Circuits are also indexed to provide fast lookups by
+// payment hash.
 //
 // TODO(andrew.shvv) make it persistent
-type circuitMap struct {
-	sync.RWMutex
-	circuits map[circuitKey]*paymentCircuit
+type CircuitMap struct {
+	mtx       sync.RWMutex
+	circuits  map[circuitKey]*PaymentCircuit
+	hashIndex map[[32]byte][]*PaymentCircuit
 }
 
-// newCircuitMap creates a new instance of the circuitMap.
-func newCircuitMap() *circuitMap {
-	return &circuitMap{
-		circuits: make(map[circuitKey]*paymentCircuit),
+// NewCircuitMap creates a new instance of the CircuitMap.
+func NewCircuitMap() *CircuitMap {
+	return &CircuitMap{
+		circuits:  make(map[circuitKey]*PaymentCircuit),
+		hashIndex: make(map[[32]byte][]*PaymentCircuit),
 	}
 }
 
-// add adds a new active payment circuit to the circuitMap.
-func (m *circuitMap) add(circuit *paymentCircuit) error {
-	m.Lock()
-	defer m.Unlock()
-
-	// Examine the circuit map to see if this circuit is already in use or
-	// not. If so, then we'll simply increment the reference count.
-	// Otherwise, we'll create a new circuit from scratch.
-	//
-	// TODO(roasbeef): include dest+src+amt in key
-	if c, ok := m.circuits[circuit.PaymentHash]; ok {
-		c.RefCount++
-		return nil
+// LookupByHTLC looks up the payment circuit by the outgoing channel and HTLC
+// IDs. Returns nil if there is no such circuit.
+func (cm *CircuitMap) LookupByHTLC(chanID lnwire.ShortChannelID, htlcID uint64) *PaymentCircuit {
+	cm.mtx.RLock()
+
+	key := circuitKey{
+		chanID: chanID,
+		htlcID: htlcID,
 	}
+	circuit := cm.circuits[key]
+
+	cm.mtx.RUnlock()
+	return circuit
+}
+
+// LookupByPaymentHash looks up and returns any payment circuits with a given
+// payment hash.
+func (cm *CircuitMap) LookupByPaymentHash(hash [32]byte) []*PaymentCircuit {
```

Comment:
Yes, it would be used when pipelining fulfills. As soon as we get a `update_fulfill` on an outgoing channel, we should immediately turn around and fulfill on all incoming channels with the same payment hash, even before the first fulfill commits.